storm-core/src/jvm/org/apache/storm/zookeeper/Zookeeper.java

```diff
@@ -0,0 +1,355 @@
+import org.apache.storm.callback.WatcherCallBack;
+import org.apache.storm.nimbus.ILeaderElector;
+import org.apache.storm.nimbus.NimbusInfo;
+import org.apache.storm.utils.Utils;
+import org.apache.storm.utils.ZookeeperAuthInfo;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.WatchedEvent;
+import org.apache.zookeeper.data.ACL;
+import org.apache.zookeeper.data.Stat;
+import org.apache.zookeeper.server.NIOServerCnxnFactory;
+import org.apache.zookeeper.server.ZooKeeperServer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.BindException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class Zookeeper {
+
+    private static Logger LOG = LoggerFactory.getLogger(Zookeeper.class);
+
+    public static CuratorFramework mkClient(Map conf, List<String> servers, Object port, String root) {
+        return mkClient(conf, servers, port, root, new DefaultWatcherCallBack());
+    }
+
+    public static CuratorFramework mkClient(Map conf, List<String> servers, Object port, Map authConf) {
+        return mkClient(conf, servers, port, "", new DefaultWatcherCallBack(), authConf);
+    }
+
+    public static CuratorFramework mkClient(Map conf, List<String> servers, Object port, String root, Map authConf) {
+        return mkClient(conf, servers, port, root, new DefaultWatcherCallBack(), authConf);
+    }
+
+    public static CuratorFramework mkClient(Map conf, List<String> servers, Object port, String root, final WatcherCallBack watcher, Map authConf) {
+        CuratorFramework fk;
+        if (authConf != null) {
+            fk = Utils.newCurator(conf, servers, port, root, new ZookeeperAuthInfo(authConf));
+        } else {
+            fk = Utils.newCurator(conf, servers, port, root);
+        }
+
+        fk.getCuratorListenable().addListener(new CuratorListener() {
+            @Override
+            public void eventReceived(CuratorFramework _fk, CuratorEvent e) throws Exception {
+                if (e.getType().equals(CuratorEventType.WATCHED)) {
+                    WatchedEvent event = e.getWatchedEvent();
+
+                    watcher.execute(event.getState(), event.getType(), event.getPath());
+                }
+
+            }
+        });
+        fk.start();
+        return fk;
+    }
+
+    /**
+     * connect ZK, register Watch/unhandle Watch
+     *
+     * @return
+     */
+    public static CuratorFramework mkClient(Map conf, List<String> servers, Object port, String root, final WatcherCallBack watcher) {
+
+        return mkClient(conf, servers, port, root, watcher, null);
+    }
+
+    public static String createNode(CuratorFramework zk, String path, byte[] data, org.apache.zookeeper.CreateMode mode, List<ACL> acls)
+            throws RuntimeException {
```

Comment:
In general we probably should not be declaring that we throw a RuntimeException.  This will force other java code to explicitly handle it.  We were wrapping all of the exceptions in Runtime because they were bubbling up through clojure to java code and were violating the java APIs, because apparently clojure can through an exception that is not declared in the API.  We probably want to file a follow on JIRA to clean up the wrapping so the Zookeeper.java APIs don't wrap exceptions, but throw what they need to, and then only when we have to we wrap the exceptions.